xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/main/java/org/xwiki/rest/internal/resources/BaseAttachmentsResource.java

```diff
@@ -348,25 +360,34 @@ protected Attachments getAttachmentsForDocument(Document doc, int start, int num
     }
 
     protected AttachmentInfo storeAndRetrieveAttachment(Document document, String attachmentName, InputStream content,
-        Boolean withPrettyNames) throws XWikiException
+        Boolean withPrettyNames) throws XWikiException, AttachmentValidationException
     {
         XWikiContext xcontext = this.xcontextProvider.get();
-        boolean alreadyExisting = document.getAttachment(attachmentName) != null;
-
-        XWikiAttachment xwikiAttachment =
-            createOrUpdateAttachment(new AttachmentReference(attachmentName, document.getDocumentReference()), content);
-        // The doc has been updated during the creation of the attachment, so we need to ensure we answer with the
-        // updated version.
-        Document updatedDoc = xwikiAttachment.getDoc().newDocument(xcontext);
-        Attachment attachment = this.modelFactory.toRestAttachment(uriInfo.getBaseUri(),
-            new com.xpn.xwiki.api.Attachment(updatedDoc, xwikiAttachment, this.xcontextProvider.get()), withPrettyNames,
-            false);
-
-        return new AttachmentInfo(attachment, alreadyExisting);
+        XWikiDocument previousDoc = xcontext.getDoc();
+        try {
+            xcontext.setDoc(document.getDocument());
+            boolean alreadyExisting = document.getAttachment(attachmentName) != null;
+
+            XWikiAttachment xwikiAttachment =
+                createOrUpdateAttachment(new AttachmentReference(attachmentName, document.getDocumentReference()),
+                    content);
+
+            // The doc has been updated during the creation of the attachment, so we need to ensure we answer with the
+            // updated version.
+            Document updatedDoc = xwikiAttachment.getDoc().newDocument(xcontext);
+            Attachment attachment = this.modelFactory.toRestAttachment(this.uriInfo.getBaseUri(),
+                new com.xpn.xwiki.api.Attachment(updatedDoc, xwikiAttachment, this.xcontextProvider.get()),
+                withPrettyNames,
+                false);
+
+            return new AttachmentInfo(attachment, alreadyExisting);
+        } finally {
```

Comment:
What's the relation of this try/finally with the new feature?

Reply:
The attachment validation requires to have the document where the attachment will be uploaded in the context. I'll add a comment.